src/components/svg_editor.tsx

```diff
@@ -0,0 +1,165 @@
+import type { Molecule } from 'openchemlib';
+import type { FormEvent, KeyboardEvent, MouseEvent } from 'react';
+import { useCallback, useMemo, useReducer, useRef } from 'react';
+
+import type { SvgRendererProps } from './svg_renderer.js';
+import { SvgRenderer } from './svg_renderer.js';
+
+export interface SvgEditorProps extends SvgRendererProps {
+  onChange: (newMolecule: Molecule) => void;
+}
+
+type State =
+  | { mode: 'view' }
+  | {
+      mode: 'atom-label-edit';
+      atomId: number;
+      formCoords: { x: number; y: number };
+    };
+const initialState: State = { mode: 'view' };
+
+type Action =
+  | {
+      type: 'startEdit';
+      atomId: number;
+      event: MouseEvent;
+    }
+  | { type: 'stopEdit' };
+
+function stateReducer(state: State, action: Action): State {
+  switch (action.type) {
+    case 'startEdit': {
+      // Ignore if we are already in edit mode
+      if (state.mode !== 'view') return state;
+
+      const { clientX, clientY } = action.event;
+      const target = action.event.target as SVGCircleElement;
+      const svg = target.closest('svg') as SVGElement;
+      const rect = svg.getBoundingClientRect();
+      const formCoords = {
+        x: clientX - rect.x,
+        y: clientY - rect.y,
+      };
+      return { mode: 'atom-label-edit', atomId: action.atomId, formCoords };
+    }
+    case 'stopEdit':
+      return { mode: 'view' };
+    default:
+      // @ts-expect-error action type narrowing
+      throw new Error(`Unknown action type: ${action.type}`);
+  }
+}
+
+/**
+ * A component that renders an SVG editor for a given molecule.
+ * Under the hood it uses the SvgRenderer component to display the molecule and catch interactions.
+ * @param props - The props for the SvgEditor component.
+ * @returns JSX.Element
+ */
+export function SvgEditor(props: SvgEditorProps) {
+  const { molecule, onChange, ...svgProps } = props;
+  const propsRef = useRef(props);
+  propsRef.current = props;
```

Comment:
I'm not sure but I remember that it's not allowed by React rules to update a ref during rendering.

Reply:
I won't block this change, but I suggest you read https://tkdodo.eu/blog/the-useless-use-callback because I agree a lot with it.
I think we have been conditioned at some point to do these kinds of ahead of time "optimizations" (useCallback and useMemo on things that don't take time to compute).